docs/design/l2-peer-protocol-spec.md

```diff
@@ -0,0 +1,507 @@
+
+#### More Precise Error Handling
+
+Beside the above generic failure reporting mechaisms the protocol contains
+context specific responses which can be useful for automatic recovery.
+
+#### `Ping`
+
+To allow for a long-lived TCP connections and to detect unresponsive peers,
+nodes can send ping messages. In this version we opt to use the `libp2p` builtin
+protocol for that: https://docs.libp2p.io/concepts/introduction/protocols/ping/
+
+This particular protocol can not be used for traffic obfuscation as it has
+constant size messages. We can revisit that decision in the future.
+
+### Handshaking
+
+#### `Init`
+
+Once a connection is established, both peers MUST send an `Init` message before
+any other messages. This is the first message revealing the features supported
+by the sending node.
+
+```haskell
+type TestnetMagic = Integer
+
+data NetworkId = Mainnet | Testnet TestnetMagic
+
+type Version = Integer
+
+type Blake2b224Hash = ByteString
+type ScriptHash = Blake2b224Hash
+
+-- maximum 32 bytes
+type TokenName = ByteString
+type Currency = (ScriptHash, TokenName)
+
+data Init = Init
+    { version :: Version
+    , networkId :: NetworkId
+    , currencies :: [Currency]
+    , routing :: Bool
+    , htlc :: Bool
+    }
+```
+
+> QUESTION: Banning:
+>
+> - Every node have full right to ban anyone.
+> - Should we allow polite banning with an Error.
+> - For sure we should allow even before the handshake to minimize resource
+>   consumption.
+
+A node:
+
+- MUST send `Init` as the first message for any connection.
+- MUST wait to receive `Init` before sending any other messages.
+- SHOULD reject if the peer doesn't support the possible version, required
+  features or network by closing he connection after sending a `Warning`.
+
+### Channel Lifecycle
+
+> Key points:
+>
+> - As it was already stated all the channel specific messages are tagged by the
+>   `channelId` from the beginning.
+> - The channel lifecycle is divided into three phases: establishing, operating
+>   and retiring.
+
+```haskell
+-- 32 bytes of some hash plus extra payload
+data ChannelId = TokenName
+
+type ChannelMessage msg = (ChannelId, m)
+```
+
+## Core Types
+
+These are core types used in different places of the protocol:
+
+```haskell
+type Positive = Integer -- Positive integer
+type Natural = Integer  -- Positive integer including 0
+type Index = Natural
+
+type Blake2b256Hash = ByteString
+
+type Amount = Natural
+
+type Milliseconds = Natural
+
+type POSIX = Milliseconds
+
+type TxId = Blake2b256Hash
+
+type TxOutRef = (TxId, Integer)
+
+type Ed25519PubKey = ByteString
+
+-- 64 bytes of Ed25519 signature
+type Signature = ByteString
+
+type Exclude = [Index]
+
+type Squash = (Amount, Index, Exclude)
+
+type Snapshot = (Squash, Squash)
+
+```
+
+### Establishing
+
+#### Overview
+
+> QUESTION(paluh): should we change how we derive chnannelId:
+>
+> - We can use txOutRef+arbitrary 32 bit integer
```

Comment:
I actually serialise the `oref` and push the byte on the end. 
There's a check on index.
And its easier in the recursion that this counts _down_ from number of tokens minted `n_mint` to 0... cos then you know when to stop.